### ISSUE TITLE
`specialize H with bindings_list` should not inline let binders

### ISSUE THREAD

**Issue by samuelgruetter** [coq/coq#11279]
I wasn't fast enough to comment on @Matafou's https://github.com/coq/coq/pull/11271 before it was merged, so I'm raising a new issue: Consider

```coq
Goal forall (a b: nat) (P: Prop),
     a = 2 * b ->
     (forall m1 m2: nat,
         let x0 := m1 * m1 in
         let x1 := x0 * x0 in
         let x2 := x1 * x1 in
         let xn := x2 * x2 in
         m1 = m2 -> xn > m2) ->
     True.
Proof.
  intros.
```

If I do 

```coq
  specialize H0 with (1 := H).
```

the type of `H0` becomes

```coq
a * a * (a * a) * 
(a * a * (a * a)) * 
(a * a * (a * a) * 
(a * a * (a * a))) > 2 * b
```

This is unfortunate for two reasons: If `specialize` is followed by further tactics which depend on this let-in structure, these will fail, and if the let-in structure is deeper than just n=3, we might get an exponentially large term size.

Desired behavior: The type of `H0` should become

```coq
let x0 := a * a in
let x1 := x0 * x0 in
let x2 := x1 * x1 in
let xn := x2 * x2 in
xn > 2 * b
```

In general, each tactic should do one thing only, and have no unexpected side-effects.

(This is a follow-up on https://github.com/coq/coq/issues/11253 and https://github.com/coq/coq/issues/9893)

Coq version: master (4a7a5c36802701d0e1b47956bb14cfc9cab99baa)

**Comment by samuelgruetter** [comment]
An alternative behavior which would also be ok for me would be to refuse to go under let binders when looking for the positions to instantiate, that is, to fail with a `No such binder` error message.

But in any case,

```coq
  specialize H0 with (m2 := 2 * b).
```

should work and not inline any let binders.

**Comment by Matafou** [comment]
It would be better indeed. I will try to avoid the expansion of Letins.

**Comment by ppedrot** [comment]
@samuelgruetter maybe I was not understood but this was precisely the content of my remark on the code. I did judge it was not a problem though, as retyping is pretty much unspecified.

**Comment by Matafou** [comment]
This is fixed by PR #11285 (which needs some code review).
Letins are reconstructed even if some variables are instantiated under them.
